src/BarrelCalorimeterImaging_geo.cpp

```diff
@@ -0,0 +1,186 @@
+
+// Detector plugin to support a hybrid central barrel calorimeter
+// The detector consists of interlayers of Pb/ScFi (segmentation in global r, phi) and W/Si (segmentation in local x, y)
+// Assembly is used as the envelope so two different detectors can be interlayered with each other
+//
+//
+// 06/19/2021: Implementation of the Sci Fiber geometry. M. Żurek
+// 07/09/2021: Support interlayers between multiple detectors. C. Peng
+// 07/23/2021: Add assemblies as mother volumes of fibers to reduce the number of daughter volumes. C. Peng, M. Żurek
+//     Reference: TGeo performance issue with large number of daughter volumes
+//     https://indico.cern.ch/event/967418/contributions/4075358/attachments/2128099/3583278/201009_shKo_dd4hep.pdf
+// 07/24/2021: Changed support implementation to avoid too many uses of boolean geometries. DAWN view seems to have
+//     issue dealing with it. C. Peng
+
+#include "DD4hep/DetFactoryHelper.h"
+#include "Math/Point2D.h"
+#include "TGeoPolygon.h"
+#include "XML/Layering.h"
+#include <functional>
+
+using namespace std;
+using namespace dd4hep;
+
+typedef ROOT::Math::XYPoint Point;
+
+// barrel ecal layers contained in an assembly
+static Ref_t create_detector(Detector& desc, xml_h e, SensitiveDetector sens)
+{
+  Layering   layering(e);
+  xml_det_t  x_det    = e;
+  int        det_id   = x_det.id();
+  string     det_name = x_det.nameStr();
+  double     offset   = x_det.attr<double>(_Unicode(offset));
+  xml_comp_t x_dim    = x_det.dimensions();
+  int        nsides   = x_dim.numsides();
+  double     inner_r  = x_dim.rmin();
+  double     dphi     = (2 * M_PI / nsides);
+  double     hphi     = dphi / 2;
+
+  DetElement sdet(det_name, det_id);
+  Volume     motherVol = desc.pickMotherVolume(sdet);
+
+  Assembly     envelope(det_name);
+  Transform3D  tr_global = Translation3D(0, 0, offset) * RotationZ(hphi);
+  PlacedVolume env_phv   = motherVol.placeVolume(envelope, tr_global);
+  sens.setType("calorimeter");
+
+  env_phv.addPhysVolID("system", det_id);
+  sdet.setPlacement(env_phv);
+
+  // build a single sector
+  DetElement sector_det("sector0", det_id);
+  Assembly   mod_vol("sector");
+
+  // keep tracking of the total thickness
+  double l_pos_z = inner_r;
+  { // =====  buildBarrelStave(desc, sens, module_volume) =====
+    // Parameters for computing the layer X dimension:
+    double tan_hphi = std::tan(hphi);
+    double l_dim_y  = x_dim.z() / 2.;
+
+    // Loop over the sets of layer elements in the detector.
+    int l_num = 1;
+    for (xml_coll_t li(x_det, _U(layer)); li; ++li) {
+      xml_comp_t x_layer         = li;
+      int        repeat          = x_layer.repeat();
+      double     l_space_between = dd4hep::getAttrOrDefault(x_layer, _Unicode(space_between), 0.);
+      double     l_space_before  = dd4hep::getAttrOrDefault(x_layer, _Unicode(space_before), 0.);
+      bool       framing         = x_layer.hasChild(_Unicode(frame));
```

Comment:
Prefer booleans with prefix e.g. `has_frame` when used in isolation on `if` statements.